### ISSUE TITLE
PSPNet Implementation - WIP

### ISSUE THREAD

**Issue by meetshah1995** [meetshah1995/pytorch-semseg#29]
@krishnavishalv is working on an implementation of the PSPNet which:

* Uses dilated version of Resnet. 
* Uses distributed batchnorm.
* Uses memory optimization as mentioned in [Training Deep Nets with Sublinear Memory Cost](https://arxiv.org/abs/1604.06174) and implemented in [mxnet-memonger](https://github.com/dmlc/mxnet-memonger)

This [chainer implementation](https://github.com/mitmul/chainer-pspnet/) by @mitmul looks a good point to start.

**Comment by meetshah1995** [comment]
@guanfuchen - I've completed the PSPNet implementation (with dilated Resnet modules but without sync BatchNorm). 

Loading pretrained models, caffemodel -> weights.npy -> load in Pytorch module is remaining.

**Comment by lucasbrynte** [comment]
@meetshah1995 This sounds great! I'm eager to try out the implementation. Is it a lot of work remaining, and is it possible to help out somehow?

**Comment by guanfuchen** [comment]
@meetshah1995  Great work! I will try it and try to do the remaining work.

**Comment by meetshah1995** [comment]
@lucasbrynte and @guanfuchen 

The only work remaining is to convert the original caffemodel weights (for Cityscapes) from [here](https://github.com/hszhao/PSPNet) to numpy format and load them in relevant modules in the pspnet class by completing this [method](https://github.com/meetshah1995/pytorch-semseg/blob/master/ptsemseg/models/pspnet.py#L47-L52).

Looking forward to your PR !

**Comment by meetshah1995** [comment]
Added load_pretrained_model method in 3dfbfb06d5827cb2edf5347b69dafff375d10d2a

Need to test with input data.

**Comment by meetshah1995** [comment]
There seems to be a bug in the implementation, every time I run the model (after loading pre-trained weights) in eval mode with a sample image, it gives different output.

**Comment by meetshah1995** [comment]
Fixed it, somewhat working  ( [out](https://i.imgur.com/g8ZPdZu.png) for [input](https://i.imgur.com/UEFyhFO.jpg) )

Maybe need to implement the tiling inference method. But for all practical purposes, we have a trainable PSPNet (pretrained load support).
